vcl/main.vcl

```diff
@@ -95,6 +80,22 @@ sub vcl_recv {
 	return(lookup);
 }
 
+sub vcl_fetch {
+
+	# Deliver stale if possible when unexpected errors are received from origin
+	# This includes 404s because Heroku will deliver a 404 if you hit the 'no app
+	# found' page.
+	if ((beresp.status >= 500 && beresp.status < 600) ||  beresp.status == 404) {
+		if (stale.exists) {
+			return(deliver_stale);
+		}
+	}
```

Comment:
This is the approach we advocate in https://docs.fastly.com/guides/performance-tuning/serving-stale-content.  Another approach might be to change `beresp.status` from 404 to 503, which should trigger varnish's default stale-if-error logic.  You could experiment with that if you want.